### ISSUE TITLE
[Python] Awaiting for `call.code()` dead-locks in the interceptor

### ISSUE THREAD

**Issue by alexeykudinkin** [grpc/grpc#34530]
### What version of gRPC and what language are you using?

1.57.0

### What operating system (Linux, Windows,...) and version?

Linux

### What runtime / compiler are you using (e.g. python version or version of gcc)

3.8

### What did you do?

While implementing a client-interceptor for Python AIO gRPC we've stumbled upon a dead-lock in the following code:

```
async def intercept_stream_unary(
        self,
        continuation: Callable[
            [ClientCallDetails, RequestType], StreamUnaryCall[RequestType, ResponseType]
        ],
        client_call_details: ClientCallDetails,
        request_iterator: AsyncIterable[RequestType],
    ) -> AsyncIterable[ResponseType]:
 
        def _emit_metrics(code: StatusCode):
           # ...

        try:
            unary_call = await continuation(client_call_details, request_iterator)
            # Following will dead-lock
            status_code = await unary_call.code()

        except AioRpcError as re:
            status_code = re.code()
            raise re
        finally: 
            _emit_metrics(status_code)

        return unary_call
```

I've traced it down to the following line leading to a dead-lock:
https://github.com/grpc/grpc/blob/90c72c0b777dddc8b48f24e2a50fe16ec5924205/src/python/grpcio/grpc/aio/_interceptor.py#L523

Essentially, when using `READER_WRITER` style (doing `stream.write(...)` calls), this would dead-lock preventing the `stream.write` call to be able to write into the queue and subsequently into the channel, before interceptors awaitable completes.

I looked at it from different angles and couldn't rationalize this particular await, so i'm thinking about removing it in a PR to address this.

### What did you expect to see?

This pattern of `await call` is working fine for other RPC types (unary, server-streaming), and is broken only for client-streaming.

### What did you see instead?

I see gRPC client-streaming dead-locking with an interceptor above.

**Comment by XuanWang-Amos** [comment]
Hi, sorry for the late reply, after some investigation, I think this is expected behavior and the proposed fix won't help.

This is what happens for AIO READER_WRITER style stream-unary call with interceptors:
* First calls `call.write()`.
* It [awaits `self._interceptors_task`](https://github.com/grpc/grpc/blob/a0c1027bb3da36bfa5d3eb2630fd62dd79d41608/src/python/grpcio/grpc/aio/_interceptor.py#L561).
* `_interceptors_task` blocks on`await unary_call.code()`.
  * `unary_call.code()` internally [`await self._cython_call.status()`](https://github.com/grpc/grpc/blob/a0c1027bb3da36bfa5d3eb2630fd62dd79d41608/src/python/grpcio/grpc/aio/_call.py#L249C64-L249C64).
  * For stream-unary call, status will be set by [this `self._set_status` call](https://github.com/grpc/grpc/blob/c19e5dcd6b4b22cc00ca61a6b836ff3edd44e16d/src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi#L473).
* And we won't be able to call `_set_status` until [`await _receive_initial_metadata` finishes](https://github.com/grpc/grpc/blob/c19e5dcd6b4b22cc00ca61a6b836ff3edd44e16d/src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi#L450).
* And last, this `_receive_initial_metadata` operation won't finish until we finish `send_receive_close` operation which [triggered by calling `done_writing()`](https://github.com/grpc/grpc/blob/a0c1027bb3da36bfa5d3eb2630fd62dd79d41608/src/python/grpcio/grpc/aio/_call.py#L509).

In short, `unary_call.code()` will be blocked until we called `done_writing()`, but since itself was triggered in an interceptor whiling calling `call.write()`, we have a deadlock. This flow is expected because we don't know for sure what status code to return until we finished the stream call by calling `done_writing()`.


If you want to access `code()`, you can try wrap `unary_call` in the interceptor and call `code()` in the wrapped call so that `_interceptors_task` won't blocked on `code()`, we'll create an example for this too.
